src/bio/bio_xstream.c

```diff
@@ -86,9 +89,337 @@ struct bio_nvme_data {
+	dprint_uint128_dec(health_page->data_units_read);
+	D_PRINT("\n");
+	D_PRINT("Data Units Written:          ");
+	dprint_uint128_dec(health_page->data_units_written);
+	D_PRINT("\n");
+	D_PRINT("Host Read Commands:          ");
+	dprint_uint128_dec(health_page->host_read_commands);
+	D_PRINT("\n");
+	D_PRINT("Host Write Commands:         ");
+	dprint_uint128_dec(health_page->host_write_commands);
+	D_PRINT("\n");
+	D_PRINT("Controller Busy Time:        ");
+	dprint_uint128_dec(health_page->controller_busy_time);
+	D_PRINT(" minutes\n");
+	D_PRINT("Power Cycles:                ");
+	dprint_uint128_dec(health_page->power_cycles);
+	D_PRINT("\n");
+	D_PRINT("Power On Hours:              ");
+	dprint_uint128_dec(health_page->power_on_hours);
+	D_PRINT(" hours\n");
+	D_PRINT("Unsafe Shutdowns:	     ");
+	dprint_uint128_dec(health_page->unsafe_shutdowns);
+	D_PRINT("\n");
+	D_PRINT("Unrecoverable Media Errors:  ");
+	dprint_uint128_dec(health_page->media_errors);
+	D_PRINT("\n");
+	D_PRINT("Lifetime Error Log Entries:  ");
+	dprint_uint128_dec(health_page->num_error_info_log_entries);
+	D_PRINT("\n");
+	D_PRINT("Warning Temperature Time:    %u minutes\n",
+		health_page->warning_temp_time);
+	D_PRINT("Critical Temperature Time:   %u minutes\n",
+		health_page->critical_temp_time);
+}
+
+/* Get the SPDK device health state log and print all useful stats */
+static void
+query_spdk_health_stats(struct bio_xs_context *ctxt, uint64_t now)
+{
+	struct spdk_bdev		*bdev = NULL;
+	struct bio_bdev_passthru_req	 pt_req;
+	int				 rc;
+	uint32_t			 numd, numdl, numdu;
+	uint32_t			 health_page_sz;
+	uint32_t			 error_page_sz;
+	uint32_t			 ctrlr_page_sz;
+
+	if (ctxt == NULL)
+		return;
+
+	/*
+	 * TODO Decide on an appropriate period to query device health
+	 * stats. Currently set at 1 min.
+	 */
+	if (ctxt->bxc_health_stat_age + DAOS_SPDK_STATS_PERIOD >= now)
+		return;
+
+	ctxt->bxc_health_stat_age = now;
+
+	if (!ctxt->bxc_health_desc)
+		return;
+
+	bdev = spdk_bdev_desc_get_bdev(ctxt->bxc_health_desc);
+	if (bdev == NULL)
+		return;
+	if (!spdk_bdev_io_type_supported(bdev, SPDK_BDEV_IO_TYPE_NVME_ADMIN)) {
+		D_ERROR("Bdev NVMe admin passthru not supported!\n");
+	}
+
+	health_page_sz = sizeof(struct spdk_nvme_health_information_page);
+	numd = health_page_sz / sizeof(uint32_t) - 1u;
+	numdl = numd & 0xFFFFu;
+	numdu = (numd >> 16) & 0xFFFFu;
+
+	memset(&pt_req, 0, sizeof(pt_req));
+	pt_req.bbr_bdev = bdev;
+	pt_req.bbr_ctxt = ctxt;
+	pt_req.bbr_channel = spdk_bdev_get_io_channel(ctxt->bxc_health_desc);
+	if (!pt_req.bbr_channel) {
+		D_ERROR("Failed to get bdev I/O channel\n");
+		return;
+	}
+	pt_req.bbr_cmd.opc = SPDK_NVME_OPC_GET_LOG_PAGE;
+	pt_req.bbr_cmd.nsid = SPDK_NVME_GLOBAL_NS_TAG;
+	pt_req.bbr_cmd.cdw10 = numdl << 16;
+	pt_req.bbr_cmd.cdw10 |= SPDK_NVME_LOG_HEALTH_INFORMATION;
+	pt_req.bbr_cmd.cdw11 = numdu;
+	pt_req.bbr_len = health_page_sz;
+	pt_req.bbr_health_buf = spdk_dma_zmalloc(pt_req.bbr_len, 0, NULL);
```

Comment:
This should be pre-allocated when the bio_blobstore is allocated.

Reply:
done